amaranth/hdl/_time.py

```diff
@@ -0,0 +1,249 @@
+        self._check_reciprocal()
+        return 1_000_000_000 / self._femtoseconds
+
+    @property
+    def gigahertz(self):
+        self._check_reciprocal()
+        return 1_000_000 / self._femtoseconds
+
+    def __lt__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return self._femtoseconds < other._femtoseconds
+
+    def __le__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return self._femtoseconds <= other._femtoseconds
+
+    def __eq__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return self._femtoseconds == other._femtoseconds
+
+    def __ne__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return self._femtoseconds != other._femtoseconds
+
+    def __gt__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return self._femtoseconds > other._femtoseconds
+
+    def __ge__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return self._femtoseconds >= other._femtoseconds
+
+    def __hash__(self):
+        return hash(self._femtoseconds)
+
+    def __bool__(self):
+        return bool(self._femtoseconds)
+
+    def __neg__(self):
+        return Period(fs = -self._femtoseconds)
+
+    def __pos__(self):
+        return self
+
+    def __abs__(self):
+        return Period(fs = abs(self._femtoseconds))
+
+    def __add__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return Period(fs = self._femtoseconds + other._femtoseconds)
+
+    def __sub__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return Period(fs = self._femtoseconds - other._femtoseconds)
+
+    def __mul__(self, other):
+        if not isinstance(other, numbers.Real):
+            return NotImplemented
+        return Period(fs = self._femtoseconds * other)
+
+    def __rmul__(self, other):
+        return self * other
+
+    def __truediv__(self, other):
+        if isinstance(other, Period):
+            return self._femtoseconds / other._femtoseconds
+        elif isinstance(other, numbers.Real):
+            return Period(fs = self._femtoseconds / other)
+        else:
+            return NotImplemented
+
+    def __floordiv__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return self._femtoseconds // other._femtoseconds
+
+    def __mod__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return Period(fs = self._femtoseconds % other._femtoseconds)
+
+    def __str__(self):
+        return self.__format__("")
+
+    def __format__(self, format_spec):
+        m = re.match(r"^([1-9]\d*)?(.\d+)?( ?)(([munpf]?)s|([kMG]?)Hz)?$", format_spec)
```

Comment:
that dot in the middle of the regex needs escaping. some of the groups near the end look like they should be non-capturing groups.

also, this is horrifying. I can parse it, but aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa